pkg/utils/Idpool.go

```diff
@@ -0,0 +1,154 @@
+#   range) to keys. IDs are assigned (or read) with get_id(key) and returned back
+#   into the pool with release_id(key). The IDPool remembers a once-assigned ID
+#   for keys so that the same ID is assigned for a key. Only when the pool runs
+#   out of unassigned keys, it will recycle released ids and assign them to new
+#   keys.
+#   Optionally, the IDPool supports reference tracking for key/ID pairs. Clients
+#   can provide a unique reference when fetching and releasing an ID for a key
+#   to support multiple independent clients.
+#   The pool will only release the ID for the key, when the last client has the
+#   released the ID with its reference. When a reference is specified in get_id()
+#   and release_id() the IDPool returns the current number of reference for the
+#   ID so that a caller knows when an ID was newly assigned (ref_count 1) or
+#   finally released (ref_count 0).
+# ---------------------------------------------------------------------------------
+*/
+type IDPool struct {
+	// self._lock = threading.Lock()
+	name         string                 // Name of pool
+	_unusedIDs   []uint32               // Yet unused IDs in pool Available ids
+	_idsInUse    map[interface{}]uint32 // Mapping key: id for currently assigned ids
+	_idsForReuse map[interface{}]uint32 // Mapping key: id for previously assigned ids
+	_refs        map[uint32][]interface{}
+	_size        int // Size of the pool
+}
+
+// IDPoolInit initialize mod ptr pool
+func IDPoolInit(name string, min uint32, max uint32) IDPool {
+	var id IDPool
+	id.name = name
+	id._unusedIDs = make([]uint32, 0)
+	for j := min; j <= (max + 1); j++ {
+		id._unusedIDs = append(id._unusedIDs, j)
+	}
+	id._size = len(id._unusedIDs)
+	id._idsInUse = make(map[interface{}]uint32)
+	id._idsForReuse = make(map[interface{}]uint32)
+	id._refs = make(map[uint32][]interface{})
+	return id
+}
+
+func (ip *IDPool) _assignID(key interface{}) uint32 {
+	// Check if there was an id assigned for that key earlier
+	id := ip._idsForReuse[key]
+	if !reflect.ValueOf(id).IsZero() {
+		// Re-use the old id
+		delete(ip._idsForReuse, key)
+	} else {
+		if len(ip._unusedIDs) != 0 {
+			// Pick an unused id
+			id = ip._unusedIDs[0]
+			ip._unusedIDs = append(ip._unusedIDs[:0], ip._unusedIDs[1:]...)
+		} else {
+			if len(ip._idsForReuse) != 0 {
+				// Pick one of the ids earlier used for another key
+				for oldKey := range ip._idsForReuse {
+					delete(ip._idsForReuse, oldKey)
+					break
+				}
+			} else {
+				// No id left
+				log.Printf("IDPool: Failed to allocate id for %+v. No free ids in pool.", key)
+				return 0
+			}
+		}
+	}
+	// Store the assigned id, if any
+	if !reflect.ValueOf(id).IsZero() {
+		ip._idsInUse[key] = id
+	}
+	return id
+}
+
+// GetID get the mod ptr id from pool
+func (ip *IDPool) GetID(key interface{}, ref interface{}) (uint32, uint32) {
+	id := ip._idsInUse[key]
+	if reflect.ValueOf(id).IsZero() {
+		// Assign a free id for the key
```

Comment:
comment is not needed, the function name states the same

Reply:
Done